### PR TITLE
DEPL-9885 WINRM_NATIVE connections cannot be established if password contains special characters

### PR DESCRIPTION
**Pull request by sameerarora**

### CODE REVIEW

File: src/main/java/com/xebialabs/overthere/util/CommandLineArgsSanitizer.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this
+ * program; if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth
+ * Floor, Boston, MA 02110-1301  USA
+ */
+package com.xebialabs.overthere.util;
+
+public class CommandLineArgsSanitizer {
```
Comment (hierynomus) [inline src/main/java/com/xebialabs/overthere/util/CommandLineArgsSanitizer.java]: Is this a generic way of escaping? If not, move it into the winrs package and rename it.

File: src/main/java/com/xebialabs/overthere/cifs/winrs/CifsWinrsConnection.java [hunk 2]
```diff
@@ -117,7 +96,7 @@ public OverthereProcess startProcess(final CmdLine cmd) {
         winrsCmd.addArgument("winrs");
         winrsCmd.addArgument("-remote:" + address + ":" + port);
         winrsCmd.addArgument("-username:" + username);
```
Comment (hierynomus) [inline src/main/java/com/xebialabs/overthere/cifs/winrs/CifsWinrsConnection.java]: Does ["Little Bobby Tables"](https://xkcd.com/327/) also get to login?

Comment (sameerarora) [inline src/main/java/com/xebialabs/overthere/cifs/winrs/CifsWinrsConnection.java]: robert& can't - let me fix that

**Comment by vpartington** [comment]
Encoding command line arguments is handled by the [CmdLineArgument](https://github.com/xebialabs/overthere/blob/master/src/main/java/com/xebialabs/overthere/CmdLineArgument.java) class. Why introduce a new one? That would solve this issue for all arguments and for all Windows connectivity issues.

IIRC @adityakalia worked on this in the context of a different bug fix. You might want to ask him.

**Comment by sameerarora** [comment]
@vpartington encodeArgumentWithSpecialCharactersForWindows method in CmdLineArgument class only puts the string in Quotes , which is not enough as the password like P@ssword+-& breaks unless we escape it with the newly added logic, specifically for winrs command case.

just spoke to @adityakalia , he suggests the logic added in xl-deploy should be moved from there to overthere as it handles many more cases.

**Comment by vpartington** [comment]
Exactly! If you move the logic that @adityakalia and his team implemented in XL Deploy to Overthere, I'm positive you'll solve this issue too.

That logic doesn't put the string in quotes like Overthere does now. IIRC, it uses the caret (^) escape that you also used in the sanitizer class you wrote.

File: src/main/java/com/xebialabs/overthere/CmdLineArgument.java [hunk 3]
```diff
@@ -117,34 +110,25 @@ protected void encodeString(String str, OperatingSystemFamily os, StringBuilder
...
-    private static boolean containsAny(String str, String chars) {
-        for (char c : chars.toCharArray()) {
-            if (str.indexOf(c) >= 0) {
-                return true;
-            }
-        }
-        return false;
-    }
-
     private static void encodeArgumentWithSpecialCharactersForWindows(String str, StringBuilder builder) {
```
Comment (hierynomus) [inline src/main/java/com/xebialabs/overthere/CmdLineArgument.java]: This now no longer calls upon the sanitizer, is that right? Or is the UNIX code wrong?

Comment (sameerarora) [inline src/main/java/com/xebialabs/overthere/CmdLineArgument.java]: Windows is already escaping with quotes and that doesn't work specifically for winrs , unix code used the logic that has been moved to the UnixSantizer

Comment (hierynomus) [inline src/main/java/com/xebialabs/overthere/CmdLineArgument.java]: But now if _I_ get confused, how do we prevent other contributors from getting equally confused. Please see whether we can simplify this.

Comment (sameerarora) [inline src/main/java/com/xebialabs/overthere/CmdLineArgument.java]: does an indication for things that are specific to winrs/winrm implementation and the generic stuff consistent between unix and windows make sense ?

**Pull request closed without merging**